### PR TITLE
SDK - Compiler - Stopped adding mlpipeline artifacts to every compiled template

### LINKED ISSUE #1421: Explicitly specify UI-metadata and Metrics artifacts for the components that really produce them
Explicitly specify UI-metadata and Metrics artifacts for the components that really produce them so that the compiler does not need to add them everywhere. 
https://github.com/kubeflow/pipelines/blob/87b96a0a7a5bec3ad502714f21d69694e46b2f4f/components/local/confusion_matrix/component.yaml#L8

### LINKED ISSUE #1422: Compiler should stop auto-adding artifacts to every op unless they're explicitly specified
Requires https://github.com/kubeflow/pipelines/issues/1421

### PR DESCRIPTION
**Pull request by Ark-kun**
Over the life of the Pipelines project we have experienced numerous issues and problems stemming from the auto-added artifacts.
This PR makes a small change to the compiler so that those artifacts are no longer auto-added to every single pipeline step. (All components that actually generate those artifacts continue to work as before.)

Fixes #1421 
Fixes #1422 
The compiler changes are in the following commit: https://github.com/kubeflow/pipelines/commit/bdd43a58adbcdbae2e607ed4593893170d51f327

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/2046)

### CODE REVIEW

File: sdk/python/kfp/dsl/_container_op.py [hunk 1]
```diff
@@ -1076,6 +1078,14 @@ def _decorated(*args, **kwargs):
                 # only proxy public callables
                 setattr(self, attr_to_proxy, _proxy(attr_to_proxy))
 
+        # Special handling for the mlpipeline-ui-metadata and mlpipeline-metrics outputs that should always be saved as artifacts
+        # TODO: Remove when outputs are always saved as artifacts
+        for output_name, path in dict(file_outputs).items():
+            normalized_output_name = re.sub('[^a-zA-Z0-9]', '-', output_name.lower())
+            if normalized_output_name in ['mlpipeline-ui-metadata', 'mlpipeline-metrics']:
+                output_artifact_paths[normalized_output_name] = path
+                del file_outputs[output_name]
```
Comment (elikatsis) [inline sdk/python/kfp/dsl/_container_op.py]: My thought is that these lines are a bit "dirty", but I get it has to deal with the existing implementation of `component`s which does not support artifacts.
Please see my review comment where I attempt to explain myself.

Comment (Ark-kun) [reply sdk/python/kfp/dsl/_container_op.py]: This part is a temporary workaround.

**Comment by Ark-kun** [comment]
> But I'd suggest putting effort on fixing this once and for all. That means:
> 
> * Step 1: Add artifact support for components
> * Step 2: Make this PR change

I have a plan for this, but it's a bit different. I want his PR to go first so that the changes are split into smaller PRs. The plan is to make make all component outputs usable as both parameters and artifacts. The component should not be responsible for deciding how its outputs will be used. Only the consumer decides how to consume the data - via a file or via a command-line argument.

**Pull request merged**